ksql-api/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java

```diff
@@ -0,0 +1,202 @@
+        .handler(this::handleCloseQuery);
+    return router;
+  }
+
+  private void handleInsertsStream(final RoutingContext routingContext) {
+    final RecordParser recordParser = RecordParser
+        .newDelimited("\n", new InsertsBodyParser(endpoints, routingContext)::handleBodyBuffer);
+    routingContext.request()
+        .handler(recordParser);
+  }
+
+  private void handleQueryStream(final RoutingContext routingContext) {
+    final HttpConnection conn = routingContext.request().connection();
+    ConnectionQueries connectionQueries = connectionsMap.get(conn);
+    if (connectionQueries == null) {
+      connectionQueries = new ConnectionQueries(conn);
+      connectionsMap.put(conn, connectionQueries);
+      conn.closeHandler(connectionQueries);
+      server.registerQueryConnection(conn);
+    }
+    final JsonObject requestBody = routingContext.getBodyAsJson();
+    final String sql = requestBody.getString("sql");
+    if (sql == null) {
+      handleError(routingContext.response(), 400, ERROR_CODE_MISSING_PARAM, "No sql in arguments");
+      return;
+    }
+    final Boolean push = requestBody.getBoolean("push");
+    if (push == null) {
+      handleError(routingContext.response(), 400, ERROR_CODE_MISSING_PARAM, "No push in arguments");
+      return;
+    }
+    final JsonObject properties = requestBody.getJsonObject("properties");
+    final QueryPublisher queryPublisher = endpoints.createQueryPublisher(sql, push, properties);
+    final QuerySubscriber querySubscriber = new QuerySubscriber(routingContext.response());
+
+    final QueryID queryID = server.registerQuery(querySubscriber);
+    connectionQueries.addQuery(queryID);
+    final JsonObject metadata = new JsonObject();
+    metadata.put("columnNames", queryPublisher.getColumnNames());
+    metadata.put("columnTypes", queryPublisher.getColumnTypes());
+    metadata.put("queryID", queryID.toString());
+    if (!push) {
+      metadata.put("rowCount", queryPublisher.getRowCount());
+    }
+    routingContext.response().write(metadata.toBuffer()).write("\n");
+    queryPublisher.subscribe(querySubscriber);
+
+    // When response is complete, publisher should be closed and query unregistered
+    routingContext.response().endHandler(v -> closeQuery(queryID, routingContext));
+  }
+
+  private boolean closeQuery(final QueryID queryID, final RoutingContext routingContext) {
+    final QuerySubscriber querySubscriber = server.removeQuery(queryID);
+    if (querySubscriber == null) {
+      return false;
+    }
+    final HttpConnection conn = routingContext.request().connection();
+    final ConnectionQueries connectionQueries = connectionsMap.get(conn);
+    connectionQueries.removeQuery(queryID);
+    querySubscriber.close();
+    return true;
+  }
+
+  private void handleCloseQuery(final RoutingContext routingContext) {
+    final JsonObject requestBody = routingContext.getBodyAsJson();
+    final String queryIDArg = requestBody.getString("queryID");
```

Comment:
again - having the API defined inline in the code as string literals makes it really hard to construct automated documentation about our public API. would be nice to have Jackson objects that we can generate documentation from

Reply:
see POJOs PR